File: src/lib/plyToSplat.ts
const SH_C0 = 0.28209479177387814;

const PLY_SIZES: Record<string, number> = {
  float: 4, float32: 4, double: 8, float64: 8,
  char: 1, uchar: 1, int8: 1, uint8: 1,
  short: 2, ushort: 2, int16: 2, uint16: 2,
  int: 4, uint: 4, int32: 4, uint32: 4,
};

export function plyToSplat(plyBuf: Buffer): Buffer {
  const END = Buffer.from('end_header\n');
  const headerEnd = plyBuf.indexOf(END);
  if (headerEnd === -1) throw new Error('Invalid PLY: no end_header');
  const dataOffset = headerEnd + END.length;
  const header = plyBuf.subarray(0, headerEnd).toString('ascii');

  const vcountMatch = header.match(/element vertex (\d+)/);
  if (!vcountMatch) throw new Error('No vertex count in PLY');
  const vcount = parseInt(vcountMatch[1]);

  const propLines = [...header.matchAll(/^property (\S+) (\S+)$/gm)];
  const propOffset: Record<string, number> = {};
  let stride = 0;
  for (const [, type, name] of propLines) {
    propOffset[name] = stride;
    stride += PLY_SIZES[type] ?? 4;
  }

  const f = (i: number, name: string) =>
    plyBuf.readFloatLE(dataOffset + i * stride + propOffset[name]);

  const x  = new Float32Array(vcount);
  const y  = new Float32Array(vcount);
  const z  = new Float32Array(vcount);
  const r  = new Float32Array(vcount);
  const g  = new Float32Array(vcount);
  const b  = new Float32Array(vcount);
  const a  = new Float32Array(vcount);
  const sx = new Float32Array(vcount);
  const sy = new Float32Array(vcount);
  const sz = new Float32Array(vcount);
  const q  = [
    new Float32Array(vcount),
    new Float32Array(vcount),
    new Float32Array(vcount),
    new Float32Array(vcount),
  ];

  for (let i = 0; i < vcount; i++) {
    x[i]  = f(i, 'x');
    y[i]  = f(i, 'y');
    z[i]  = f(i, 'z');
    r[i]  = Math.min(1, Math.max(0, 0.5 + SH_C0 * f(i, 'f_dc_0')));
    g[i]  = Math.min(1, Math.max(0, 0.5 + SH_C0 * f(i, 'f_dc_1')));
    b[i]  = Math.min(1, Math.max(0, 0.5 + SH_C0 * f(i, 'f_dc_2')));
    a[i]  = 1.0 / (1.0 + Math.exp(-f(i, 'opacity')));
    sx[i] = Math.exp(f(i, 'scale_0'));
    sy[i] = Math.exp(f(i, 'scale_1'));
    sz[i] = Math.exp(f(i, 'scale_2'));
    const q0 = f(i, 'rot_0'), q1 = f(i, 'rot_1'), q2 = f(i, 'rot_2'), q3 = f(i, 'rot_3');
    const qlen = Math.max(1e-8, Math.sqrt(q0*q0 + q1*q1 + q2*q2 + q3*q3));
    q[0][i] = q0 / qlen;
    q[1][i] = q1 / qlen;
    q[2][i] = q2 / qlen;
    q[3][i] = q3 / qlen;
  }

  const order = Array.from({ length: vcount }, (_, i) => i).sort((a2, b2) => a[b2] - a[a2]);

  const out = Buffer.allocUnsafe(vcount * 32);
  for (let i = 0; i < vcount; i++) {
    const j = order[i];
    const off = i * 32;
    out.writeFloatLE(x[j],  off);
    out.writeFloatLE(y[j],  off + 4);
    out.writeFloatLE(z[j],  off + 8);
    out.writeFloatLE(sx[j], off + 12);
    out.writeFloatLE(sy[j], off + 16);
    out.writeFloatLE(sz[j], off + 20);
    out.writeUInt8(Math.round(r[j] * 255), off + 24);
    out.writeUInt8(Math.round(g[j] * 255), off + 25);
    out.writeUInt8(Math.round(b[j] * 255), off + 26);
    out.writeUInt8(Math.round(a[j] * 255), off + 27);
    out.writeUInt8(Math.min(255, Math.max(0, Math.round(q[0][j] * 128 + 128))), off + 28);
    out.writeUInt8(Math.min(255, Math.max(0, Math.round(q[1][j] * 128 + 128))), off + 29);
    out.writeUInt8(Math.min(255, Math.max(0, Math.round(q[2][j] * 128 + 128))), off + 30);
    out.writeUInt8(Math.min(255, Math.max(0, Math.round(q[3][j] * 128 + 128))), off + 31);
  }

  return out;
}

File: scripts/ply_to_splat.py
"""
Convert a 3DGS PLY file to the .splat binary format expected by @react-three/drei's Splat component.

.splat record layout (32 bytes each):
  [0..11]  position: 3 x float32 (x, y, z)
  [12..23] scale:    3 x float32 (exp(scale_0), exp(scale_1), exp(scale_2))
  [24..27] color:    4 x uint8  (r, g, b, opacity)
  [28..31] rotation: 4 x uint8  (rot_0..3 as (val*128+128) clamped to [0,255])

Matches the TypeScript plyToSplat in src/app/api/facelift/route.ts exactly:
  - Sorts splats by descending opacity (required for correct alpha blending)
  - Normalizes quaternions before packing
  - SH DC → linear color: channel = 0.5 + SH_C0 * f_dc_i
  - Opacity: sigmoid(raw_opacity)
"""

import struct, math, sys
from pathlib import Path

SH_C0 = 0.28209479177387814

def sigmoid(x):
    return 1.0 / (1.0 + math.exp(-x))

def float_to_u8(v):
    return max(0, min(255, int(round(v * 255))))

def rot_to_u8(v):
    return max(0, min(255, int(round(v * 128 + 128))))

def convert(src: Path, dst: Path):
    with open(src, 'rb') as f:
        header_bytes = b''
        while True:
            line = f.readline()
            header_bytes += line
            if b'end_header' in line:
                break
        header = header_bytes.decode('utf-8', errors='replace')

        props = []
        for ln in header.splitlines():
            if ln.startswith('property float '):
                props.append(ln.split()[-1])

        n_verts = int(next(
            ln.split()[-1] for ln in header.splitlines() if ln.startswith('element vertex')
        ))
        stride = len(props) * 4

        idx = {name: i for i, name in enumerate(props)}
        required = ['x','y','z','scale_0','scale_1','scale_2','f_dc_0','f_dc_1','f_dc_2','opacity','rot_0','rot_1','rot_2','rot_3']
        for r in required:
            if r not in idx:
                raise ValueError(f"Missing property: {r}")

        data = f.read(n_verts * stride)

    fmt = '<' + 'f' * len(props)

    # Parse all vertices
    splats = []
    for i in range(n_verts):
        vals = struct.unpack_from(fmt, data, i * stride)
        x  = vals[idx['x']]
        y  = vals[idx['y']]
        z  = vals[idx['z']]
        sx = math.exp(vals[idx['scale_0']])
        sy = math.exp(vals[idx['scale_1']])
        sz = math.exp(vals[idx['scale_2']])
        r  = float_to_u8(max(0.0, 0.5 + SH_C0 * vals[idx['f_dc_0']]))
        g  = float_to_u8(max(0.0, 0.5 + SH_C0 * vals[idx['f_dc_1']]))
        b  = float_to_u8(max(0.0, 0.5 + SH_C0 * vals[idx['f_dc_2']]))
        a  = sigmoid(vals[idx['opacity']])

        # Normalize quaternion (matches TypeScript)
        q0r = vals[idx['rot_0']]
        q1r = vals[idx['rot_1']]
        q2r = vals[idx['rot_2']]
        q3r = vals[idx['rot_3']]
        qlen = max(1e-8, math.sqrt(q0r*q0r + q1r*q1r + q2r*q2r + q3r*q3r))
        q0 = rot_to_u8(q0r / qlen)
        q1 = rot_to_u8(q1r / qlen)
        q2 = rot_to_u8(q2r / qlen)
        q3 = rot_to_u8(q3r / qlen)

        splats.append((x, y, z, sx, sy, sz, r, g, b, a, q0, q1, q2, q3))

    # Sort by descending opacity (matches TypeScript)
    splats.sort(key=lambda s: s[9], reverse=True)

    out = bytearray(n_verts * 32)
    for i, (x, y, z, sx, sy, sz, r, g, b, a, q0, q1, q2, q3) in enumerate(splats):
        base = i * 32
        struct.pack_into('<fff', out, base,      x, y, z)
        struct.pack_into('<fff', out, base + 12, sx, sy, sz)
        struct.pack_into('4B',  out, base + 24, r, g, b, float_to_u8(a))
        struct.pack_into('4B',  out, base + 28, q0, q1, q2, q3)

    dst.write_bytes(out)
    print(f"Wrote {n_verts} splats → {dst}  ({dst.stat().st_size / 1024 / 1024:.1f} MB)")

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print("Usage: python ply_to_splat.py input.ply output.splat")
        sys.exit(1)
    convert(Path(sys.argv[1]), Path(sys.argv[2]))
